server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -997,6 +1011,51 @@ public void testClusterCannotFormWithFailingJoinValidation() {
         assertTrue(cluster.clusterNodes.stream().allMatch(cn -> cn.getLastAppliedClusterState().version() == 0));
     }
 
+    public void testCannotJoinClusterWithDifferentUUID() throws IllegalAccessException {
+        final Cluster cluster1 = new Cluster(randomIntBetween(1, 3));
+        cluster1.runRandomly();
+        cluster1.stabilise();
+
+        final Cluster cluster2 = new Cluster(3);
+        cluster2.runRandomly();
+        cluster2.stabilise();
+
+        final ClusterNode shiftedNode = randomFrom(cluster2.clusterNodes).restartedNode();
+        final ClusterNode newNode = cluster1.new ClusterNode(nextNodeIndex.getAndIncrement(),
+            shiftedNode.getLocalNode(), n -> shiftedNode.persistedState);
+        cluster1.clusterNodes.add(newNode);
+
+        MockLogAppender mockAppender = new MockLogAppender();
+        mockAppender.start();
+        mockAppender.addExpectation(
+            new MockLogAppender.SeenEventExpectation(
+                "test1",
+                JoinHelper.class.getCanonicalName(),
+                Level.INFO,
+                "*failed to join*"));
+        Logger joinLogger = LogManager.getLogger(JoinHelper.class);
+        Loggers.addAppender(joinLogger, mockAppender);
+        cluster1.runFor(10000, "failing join validation");
+        try {
+            mockAppender.assertAllExpectationsMatched();
+        } finally {
+            Loggers.removeAppender(joinLogger, mockAppender);
+            mockAppender.stop();
+        }
+        assertTrue(newNode.getLastAppliedClusterState().version() == 0);
+
+        // reset clusterUUIDCommitted (and node / cluster state term) to let node join again
+        final ClusterNode detachedNode = newNode.restartedNode(
+            metaData -> MetaData.builder(metaData)
+                .clusterUUIDCommitted(false)
+                .coordinationMetaData(CoordinationMetaData.builder(metaData.coordinationMetaData())
+                    .term(0L).build())
+                .build(),
+            term -> 0L);
```

Comment:
Seems that resetting `term` and `currentTerm` to 0 was required? shall we do the same in elasticsearch-node tool? What about version?

Reply:
yes, this was required. I'll leave the definitive way to the tool (added TODO in c3e0aa2572b)